src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs

```diff
@@ -0,0 +1,582 @@
+            /// LocalDeclarationSyntaxNode of x will the <see cref="StatementContainingCallee"/>.
+            /// And if there is any statements needs inserted, it needs to be inserted before this node.
+            /// </summary>
+            public SyntaxNode StatementContainingCallee { get; }
+
+            /// <summary>
+            /// Inline content for the callee method. It should replace <see cref="SyntaxNodeToReplace"/>.
+            /// </summary>
+            public SyntaxNode InlineSyntaxNode { get; }
+
+            /// <summary>
+            /// SyntaxNode needs to be replaced by <see cref="InlineSyntaxNode"/>
+            /// </summary>
+            public SyntaxNode SyntaxNodeToReplace { get; }
+
+            /// <summary>
+            /// Indicate is <see cref="InlineSyntaxNode"/> has Await Expression.
+            /// </summary>
+            public bool ContainsAwaitExpression { get; }
+
+            /// <summary>
+            /// A preferred name used to generated a declaration when the
+            /// inline method has a return value but is not assigned to a variable.
+            /// Example:
+            /// void Caller()
+            /// {
+            ///     Callee();
+            /// }
+            /// int Callee()
+            /// {
+            ///     return 1;
+            /// };
+            /// After it should be:
+            /// void Caller()
+            /// {
+            ///     int temp = 1;
+            /// }
+            /// int Callee()
+            /// {
+            ///     return 1;
+            /// };
+            /// </summary>
+            private const string TemporaryName = "temp";
+
+            private InlineMethodContext(
+                ImmutableArray<SyntaxNode> statementsToInsertBeforeCallee,
+                SyntaxNode statementContainingCallee,
+                SyntaxNode inlineSyntaxNode,
+                SyntaxNode syntaxNodeToReplace,
+                bool containsAwaitExpression)
+            {
+                StatementsToInsertBeforeCallee = statementsToInsertBeforeCallee;
+                StatementContainingCallee = statementContainingCallee;
+                InlineSyntaxNode = inlineSyntaxNode;
+                SyntaxNodeToReplace = syntaxNodeToReplace;
+                ContainsAwaitExpression = containsAwaitExpression;
+            }
+
+            public static async Task<InlineMethodContext> GetInlineContextAsync(
+                AbstractInlineMethodRefactoringProvider<TInvocationNode, TExpression, TArgumentSyntax, TMethodDeclarationSyntax, TIdentifierName> inlineMethodRefactoringProvider,
+                Document document,
+                SyntaxNode calleeInvocationNode,
+                IMethodSymbol calleeMethodSymbol,
+                TMethodDeclarationSyntax calleeMethodDeclarationNode,
+                SyntaxNode statementContainsCallee,
+                MethodParametersInfo methodParametersInfo,
+                CancellationToken cancellationToken)
+            {
```

Comment:
why is this a static method here that takes the provider, instead of just being an instance method on the provider itself?

Reply:
I was thinking the provider only needs to do some document side changes (like replace, remove..) based on the context so put all the analysis code here to here (a factory method) is might be clearer. 
But there is no absolutely reason that it is needed to be here. If you think as an instance method it will be easy to understand then I can modify it :)

Reply:
:) I think I get the reason that's because the generics list is so long and it looks like a disaster.

Reply:
a static method that takes in an object and accesses a bunch of private state on it feels like a major smell.  If it's just an instance method it wil lbe simpler and cleaner IMO :)